### PR TITLE
[RSDK-9216] - allocate buffer for response message per RPC call

### PR DESCRIPTION
**Pull request by gvaradarajan**

### CODE REVIEW

**Comment by gvaradarajan** [comment]
Camera code is untested on hardware, will refrain from merging until it is but wanted the code review regardless

**Comment by mattjperez** [comment]
tested with fake camera and esp32camera, still works

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [inline micro-rdk/src/common/grpc.rs:1359]: shouldn't this use the new allocator?

File: micro-rdk/src/common/grpc.rs
Comment (mattjperez) [inline micro-rdk/src/common/grpc.rs:1386]: same here

File: micro-rdk/src/esp32/rpc_heap.rs
Comment (mattjperez) [inline micro-rdk/src/esp32/rpc_heap.rs:22]: I think these deserve their own error types and get propagated upward to `GrpcError`

File: micro-rdk-server/native/main.rs
Comment (gvaradarajan) [reply micro-rdk-server/native/main.rs:86]: I don't quite see a reason to?

File: micro-rdk/src/common/grpc.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/grpc.rs:1359]: I didn't bother because it would simply change immediately in the next PR

File: micro-rdk/src/esp32/rpc_heap.rs
Comment (gvaradarajan) [reply micro-rdk/src/esp32/rpc_heap.rs:22]: I'd prefer to save doing that until we have an implementation of the trait that would actually want to surface any other error

File: micro-rdk-server/native/main.rs
Comment (mattjperez) [reply micro-rdk-server/native/main.rs:86]: just that the leap from `BytesMut` to `NativeHeapAllocation` feels odd. One is external with a generic name, not hinting on how it might work underneath, the other is the opposite (granted, because we're the ones implementing it)

**Review by mattjperez — APPROVED** [review]
LGTM, but I think @acmorrow would like to coordinate this with his PR

**Comment by acmorrow** [comment]
@gvaradarajan - Yes, please don't merge this until we land #347.

**Comment by acmorrow** [comment]
> @gvaradarajan - Yes, please don't merge this until we land #347.

Also, please let me get a round of review in on it.

File: micro-rdk-server/native/main.rs
Comment (gvaradarajan) [reply micro-rdk-server/native/main.rs:86]: Ok, that's a fair point about clarity, I'll make this change

File: micro-rdk/src/esp32/rpc_heap.rs
Comment (npmenard) [inline micro-rdk/src/esp32/rpc_heap.rs:27]: you shouldn't set the len it implies that the underlying data is initialized

File: micro-rdk/src/esp32/rpc_heap.rs
Comment (npmenard) [inline micro-rdk/src/esp32/rpc_heap.rs:44]: the Esp32RpcHeapAllocation object will be dropped here and memory will be release.

File: micro-rdk-server/esp32/main.rs
Comment (npmenard) [inline micro-rdk-server/esp32/main.rs:156]: I don't think we should add a Generic here

**Comment by gvaradarajan** [comment]
> I am not sure the code works the way you intended it to. Perhaps we should use the allocator_api and replace all Bytes with Vec<u8, AllocatorESP32> since it implements Buf + BufMut that should be possible. To simplify the code I would use a global type so Vec<u8, A> get swapped for Vec in native builds

@npmenard, we can't use the allocator API because it panics rather than returns an error when unable to allocate (except in Nightly Rust)

File: micro-rdk/src/esp32/rpc_heap.rs
Comment (gvaradarajan) [reply micro-rdk/src/esp32/rpc_heap.rs:44]: isn't that good?

Comment (npmenard) [reply micro-rdk/src/esp32/rpc_heap.rs:44]: yes but it doesn't look like we ever use the memory allocated by get_allocation.

**Comment by npmenard** [comment]
esp toolchain should be nightly no?

File: micro-rdk-server/esp32/main.rs
Comment (mattjperez) [reply micro-rdk-server/esp32/main.rs:156]: should it be an input parameter to `run_forever`?

**Comment by gvaradarajan** [comment]
> esp toolchain should be nightly no?

yes but not for native builds. So making a consistent API requires wrapping

**Comment by gvaradarajan** [comment]
This PR has a lot of fundamental problems, closing

**Pull request closed without merging**
